README.md
# weatherkit-rust

A Rust CLI program to print current conditions and daily/hourly forecast to the console.

Please read [`authorization.md`](authorization.md) as you need an Apple Developer Program membership and four environment variables set up to use this CLI tool. These are said env vars:

- `WEATHERKIT_KEY_ID`
- `WEATHERKIT_SERVICE_ID`
- `WEATHERKIT_TEAM_ID`
- `WEATHERKIT_KEY_PATH`

and the aforementioned document explains how to get each of those values from Apple.

ALSO NOTE that the program defaults to my location. Make sure to run it with params.

```
boB Rudis (@hrbrmstr)
Print Apple WeatherKit REST API weather conditions and hourly/daily foreacast to the console.

USAGE:
    weatherkit [OPTIONS]

OPTIONS:
    -h, --help           Print help information
        --lang <LANG>    [default: en]
        --lat <LAT>      [default: 43.2683199]
        --lon <LON>      [default: -70.8635506]
    -V, --version        Print version information
```

You get something like this:

![](sample-forecast.png)

## Building

```
git clone git@github.com:hrbrmstr/weatherkit-rust
cargo build --release 
```

## Installing

Homebrew users can do:

```bash
brew install hrbrmstr/tap/weatherkit
```

The following will put:

- `weatherkit`

into `~/.cargo/bin` unless you've modified the behaviour of `cargo install`.

```
$ cargo install --git https://github.com/hrbrmstr/weatherkit-rust
```


CHANGELOG.md
# Changelog

## 0.3.4

* clean up rust-analyzer warnings and update all dependencies

## 0.3.3

* negative numbers ([457f82e](https://github.com/hrbrmstr/weatherkit-rust/commit/457f82e8f540afa4f7adb29e40e974e3b411d18f))

## 0.3.1

- fix negative values in cmdline values

## [0.3.1](https://github.com/hrbrmstr/weatherkit-rust/compare/v0.3.0...v0.3.1) (2023-08-21)


### Bug Fixes

### Features

it's kind of actually usable now? fixed the problem with using negative values on the cli


## [0.3.0](https://github.com/hrbrmstr/weatherkit-rust/compare/v0.2.3...v0.3.0) (2022-07-31)


### Features

* Weather alerts! ([e47bfbe](https://github.com/hrbrmstr/weatherkit-rust/commit/e47bfbe3c207ddc99ab76a29dffad96a965e2ffb))
* Weather alerts! ([b690eb6](https://github.com/hrbrmstr/weatherkit-rust/commit/b690eb6d6bff8c252cd53818c4a3f2d7e9a02b36))

## [0.2.3](https://github.com/hrbrmstr/weatherkit-rust/compare/v0.2.2...v0.2.3) (2022-07-20)


### Bug Fixes

* Missed some or there are new enumerations from Apple ([4becc32](https://github.com/hrbrmstr/weatherkit-rust/commit/4becc32c6831560f8d75ce98d8900a0b61d767ab))

## [0.2.2](https://github.com/hrbrmstr/weatherkit-rust/compare/v0.2.1...v0.2.2) (2022-07-17)


### Bug Fixes

* missed an enum ([6352261](https://github.com/hrbrmstr/weatherkit-rust/commit/6352261cedecfdca464364a9ee536fe3493e9b1a))

## [0.2.1](https://github.com/hrbrmstr/weatherkit-rust/compare/v0.2.0...v0.2.1) (2022-07-17)


### Bug Fixes

* handful of JSON elements needed to be Option-al ([d8994c1](https://github.com/hrbrmstr/weatherkit-rust/commit/d8994c1b14dad03634645bb74625d5d2da879490))

## [0.2.0](https://github.com/hrbrmstr/weatherkit-rust/compare/v0.1.0...v0.2.0) (2022-07-12)


### Features

* placename lookup ([c35fad4](https://github.com/hrbrmstr/weatherkit-rust/commit/c35fad4b3b046fa73af663eab346bbd3ef3c2e72))

## 0.1.0 (2022-07-10)


### Features

* automatic binary releases (hopefully) ([1562bfa](https://github.com/hrbrmstr/weatherkit-rust/commit/1562bfa8855affd182f0ca186c20f7bc8120a544))


### weatherkit

* release 0.1.0 ([f15a517](https://github.com/hrbrmstr/weatherkit-rust/commit/f15a5170f05267a9c3671b2294a1898cdad366a4))
* release 0.1.0 ([803a7bc](https://github.com/hrbrmstr/weatherkit-rust/commit/803a7bc65571520e0fdf134fab047d5405b96493))
* release 0.1.0 ([8c964a2](https://github.com/hrbrmstr/weatherkit-rust/commit/8c964a20531bff2162035494e46c8ae4118e4291))
* release 0.1.0 ([37e3e7c](https://github.com/hrbrmstr/weatherkit-rust/commit/37e3e7cd720dab1e12e72e5784957fe12bd20a9f))


src/main.rs
// #[allow(unused_imports)]
// use log::{info, trace, warn};

use std::time::{UNIX_EPOCH, SystemTime};
use std::io::Read;
use std::io::BufReader;
use std::fs::File;
use std::env;
use std::collections::HashSet;

use serde_json::json;

use jsonwebtokens as jwt;
use jwt::{Algorithm, AlgorithmID, encode};

use tz_search::lookup;

use format_num::NumberFormat;

use pad::{PadStr, Alignment};

use chrono::prelude::*;
use chrono::{DateTime, Utc, SecondsFormat};
use chrono_tz::Tz;

extern crate colored;
use colored::{*};

use clap::Parser;

mod weatherkitweather;
use weatherkitweather::{*};

mod placename;
use placename::{*};

mod utils;
use utils::{*};

/// Mandatory Apple Weather trademark. The Apple glyph is a pain
/// to type so this makes it a bit easier to handle
const APPLE_WEATHER_TRADEMARK: &str = " Weather";

#[derive(Parser, Debug)]
#[clap(author, version, about, long_about = None)]
struct Args {

  // latitude
  #[clap(long, allow_hyphen_values = true, default_value_t = String::from("43.2683199"))]
  lat: String,

  // longitude
  #[clap(long, allow_hyphen_values = true, default_value_t = String::from("-70.8635506"))]
  lon: String,

  // language
  #[clap(long, default_value_t = String::from("en"))]
  lang: String,

  #[clap(long, default_value_t = String::from("US"))]
  iso2c: String,
  
}

/// Setup the JSON Web Token from the required environment variables.
/// 
/// # Arguments
/// 
/// * None
/// 
/// # Required Environment Variables
/// 
/// - `WEATHERKIT_KEY_ID`
/// - `WEATHERKIT_SERVICE_ID`
/// - `WEATHERKIT_TEAM_ID`
/// - `WEATHERKIT_KEY_PATH`
/// 
/// See <https://github.com/hrbrmstr/weatherkit-rust/blob/batman/authorization.md> for how to 
/// set up those values.
fn setup_jwt() -> String {
  
  let wxkit_keyid = env::var("WEATHERKIT_KEY_ID").expect("Please set WEATHERKIT_KEY_ID");
  let wxkit_service_id = env::var("WEATHERKIT_SERVICE_ID").expect("Please set WEATHERKIT_SERVICE_ID");
  let wxkit_teamid = env::var("WEATHERKIT_TEAM_ID").expect("Please set WEATHERKIT_TEAM_ID");
  let wxkit_key_path = env::var("WEATHERKIT_KEY_PATH").expect("Please set WEATHERKIT_KEY_PATH");
  
  let f = File::open(wxkit_key_path.as_str()).expect("Error opening key file at WEATHERKIT_KEY_PATH");
  let mut reader = BufReader::new(f);
  let mut buffer = Vec::new();    
  reader.read_to_end(&mut buffer).expect("Error reading key file at WEATHERKIT_KEY_PATH");
  let key_buf: &[u8] = &buffer;
  
  let alg = Algorithm::new_ecdsa_pem_signer(AlgorithmID::ES256, key_buf).expect("Error decoding PEM");
  
  let now = SystemTime::now().duration_since(UNIX_EPOCH).expect("Timey wimey").as_secs();
  let expires = now + 3600;
  
  let claims = json!({
    "iss": wxkit_teamid,
    "iat": now,
    "exp": expires,
    "sub": wxkit_service_id
  });
  
  let header = json!({
    "alg": "ES256",
    "kid": wxkit_keyid,
    "id": format!("{}.{}", wxkit_teamid, wxkit_service_id)
  });
  
  encode(&header, &claims, &alg).expect("Error creating JWT")

}

/// Retrieve weather data from Apple's WeatherKit REST API
/// 
/// # Arguments
/// 
/// - `token` - JWT from [setup_jwt()] call
/// - `latitude` - latitude
/// - `longitude` - longitude
/// - `language` - language code
/// - `tzone` - time zone string
/// - `utc_now` - "now" in UTC
fn get_weatherkit_weather(token: String, 
                          latitude: String,
                          longitude: String, 
                          language: String, 
                          iso2c: String,
                          tzone: String, 
                          utc_now: DateTime<Utc>) -> 
                          weatherkitweather::WeatherKitWeather {
  
  let utc_now_fmt = utc_now.to_rfc3339_opts(SecondsFormat::Secs, true);

  let url = format!(
    "https://weatherkit.apple.com/api/v1/weather/{}/{}/{}?timezone={}&country={}&countryCode={}&dataSets=currentWeather,forecastDaily,forecastHourly,weatherAlerts&dailyStart={}&hourlyStart={}", 
    language, latitude, longitude, tzone, iso2c, iso2c, utc_now_fmt, utc_now_fmt
  );
  
  let client = reqwest::blocking::Client::new();
  
  
  
  client
    .get(url)
    .header("Authorization", format!("Bearer {}", token))
    .header("Accept", "application/json")
    .send().expect("Error retrieving weather from WeatherKit REST API")
    .json::<WeatherKitWeather>()
    .expect("Error retrieving weather info.")

}

/// Print everything
/// 
/// Setup the JWT, parse the cmdline args, estimate timezone, get the weather, print all the things.
fn main() {

  // simple_logger::SimpleLogger::new().env().init().unwrap();

  let token = setup_jwt();

  let args = Args::parse();

  let iso2c: String = args.iso2c;

  let lat = args.lat;
  let lon = args.lon;
  
  let place = get_placename(lat.parse().unwrap(), lon.parse().unwrap());
  
  let tzone = lookup(lat.parse().unwrap(), lon.parse().unwrap()).unwrap();
  let tz: Tz = tzone.parse().unwrap();

  let utc_now = Utc::now();

  let resp = get_weatherkit_weather(token, lat, lon, args.lang, iso2c, tzone, utc_now);
  
  let num = NumberFormat::new();

  let as_of_time = DateTime::parse_from_rfc3339(resp.current_weather.as_of.as_str()).unwrap().with_timezone(&tz);
  
  // Attribution labeling required by Apple
  // println!("{} daily forecast for ({}, {}) as of {}\n", APPLE_WEATHER_TRADEMARK, resp.current_weather.metadata.latitude, resp.current_weather.metadata.longitude, as_of_time);
  println!("{} daily forecast for {} as of {}\n", APPLE_WEATHER_TRADEMARK, place, as_of_time);
  
  // TODO metric option for everything that's displayed

  println!(" Conditions: {}",         format_args!("{}", condition_code(&resp.current_weather.condition_code)));
  println!("Temperature: {}°F",       num.format(".1f", c_to_f(resp.current_weather.temperature)));
  println!(" Feels like: {}°F",       num.format(".1f", c_to_f(resp.current_weather.temperature_apparent)));
  println!("  Dew Point: {}°F",       num.format(".1f", c_to_f(resp.current_weather.temperature_dew_point)));
  println!("       Wind: {} mph",     num.format(".0f", kmph_to_mph(resp.current_weather.wind_speed)));
  println!("   Pressure: {} mb ({})", num.format(".0f", resp.current_weather.pressure), format_args!("{:?}", resp.current_weather.pressure_trend));
  println!(" Visibility: {} miles",   meters_to_miles(resp.current_weather.visibility));
  println!("   UV Index: {} {}",      uv_label(resp.current_weather.uv_index, true), uv_label(resp.current_weather.uv_index, false));
  println!();

  let hrs = resp.forecast_hourly.hours;
  
  // get max lengths of various values so we can pad properly and not too much

  let max_hr_temp_len = hrs[0..23].iter().fold(0, |accum, hr| {
    let hr_len = num.format(".0f", c_to_f(hr.temperature)).len();
    if accum >= hr_len { accum } else { hr_len }
  });
  
  let max_hr_humid_len = hrs[0..23].iter().fold(0, |accum, hr| {
    let hr_len = num.format(".0f", hr.humidity * 100.0).len();
    if accum >= hr_len { accum } else { hr_len }
  });

  let max_hr_condiiton_len = hrs[0..23].iter().fold(0, |accum, hr| {
    let hr_len = condition_code(&hr.condition_code).len();
    if accum >= hr_len { accum } else { hr_len }
  });
   
  // keep track of day's we've seen & printed so we don't print them more than once
  let mut day_set: HashSet<String> = HashSet::new();

  // for idx in 0..hrs[0..23].len() {
  for hr in hrs {

    let hour = hr;
  
    let utc_fcast_time = DateTime::parse_from_rfc3339(hour.forecast_start.as_str()).unwrap();
    let local_time = utc_fcast_time.with_timezone(&tz);
    let weekday = format!("{}", local_time.weekday()).pad_to_width_with_alignment(5, Alignment::Right);
    let printed_str = if local_time.date_naive() == utc_now.date_naive() { "Today" } else { weekday.as_str() };
    
        println!(
      "{} @ {}:00 │ 🌡  {}°F │ 💦 {}% │ {} mb {} │ {} │ {} │ {}",
      if day_set.contains(printed_str) { "     " } else { printed_str },
      num.format("02d", local_time.hour()),
      num.format(".0f", c_to_f(hour.temperature)).pad_to_width_with_alignment(max_hr_temp_len, Alignment::Right),
      num.format(".0f", hour.humidity * 100.0).pad_to_width_with_alignment(max_hr_humid_len, Alignment::Right),
      num.format("4.0f", hour.pressure),
      match &hour.pressure_trend {
        Some(trend) => pressure_trend(trend),
        None => " ".to_string()
      },
      precip_type(&hour.precipitation_type, hour.daylight.unwrap_or(true)),
      condition_code(&hour.condition_code).pad_to_width(max_hr_condiiton_len),
      uv_label(hour.uv_index, true)
    );

    day_set.insert(printed_str.to_string());
    
  }
  
  println!();

  let days = resp.forecast_daily.days;

  // get all the info we need to rescale the temp ranges to 1..30 (# chars in the fake bars)

  let min_temps: Vec<i64> = days.iter().map(|d| c_to_f(d.temperature_min) as i64).collect();
  let max_temps: Vec<i64> = days.iter().map(|d| c_to_f(d.temperature_max) as i64).collect();

  let min_temp = min_temps.iter().min().expect("INCOSPIGLE!");
  let max_temp = max_temps.iter().max().expect("INCOSPIGLE!");

  let max_day_temp_len = days.iter().fold(0, |accum, day| {
    let day_len = num.format("d", c_to_f(day.temperature_max) as i32).len();
    if accum >= day_len { accum } else { day_len }
  });

  let block = "▆";

  for d in days {

    let day = d;

    let utc_fcast_time = DateTime::parse_from_rfc3339(day.forecast_start.as_str()).unwrap();
    let local_time = utc_fcast_time.with_timezone(&tz);
    let weekday = format!("{}", local_time.weekday()).pad_to_width_with_alignment(5, Alignment::Right);
    let weekday_str = if local_time.date_naive() == utc_now.date_naive() { "Today" } else { weekday.as_str() };

    let day_min = c_to_f(day.temperature_min) as i32; // num.format doesn't like i64
    let day_max = c_to_f(day.temperature_max) as i32;
    let day_min_scaled = rescale_val(day_min as i64, *min_temp, *max_temp, 1, 30);
    let day_max_scaled = rescale_val(day_max as i64, *min_temp, *max_temp, 1, 30);

    print!(
      "{} {}°F ", 
      weekday_str,
      num.format("d", day_min).pad_to_width_with_alignment(max_day_temp_len, Alignment::Right),
    );

    for cell in 1..30 {
      if cell >= day_min_scaled && cell <= day_max_scaled {
        print!("{}", block.color("yellow"));
      } else {
        print!("{}", block.color("bright black"));
      }
    }
    
    println!(
      " {}°F {} {}",
      num.format("d", day_max).pad_to_width_with_alignment(max_day_temp_len, Alignment::Left),
      uv_label(day.max_uv_index, true),
      condition_code(&day.condition_code)
    );

  }

  println!();

  if let Some(alerts) = resp.weather_alerts {

      let alerts = alerts.alerts;
      let num_alerts = alerts.len();

      if num_alerts > 0 {
        let is_are = if num_alerts == 1 { "IS" } else { "ARE" };
        let s = if num_alerts == 1 { "" } else { "S" };

        println!("🚨 THERE {} {} ACTIVE WEATHER ALERT{}\n",is_are, num_alerts, s);

        for alert in alerts {

          println!("     Area: {}", alert.area_name);
          println!("     Kind: {}", alert.description);
          println!(" Severity: {}", severity_trans(alert.severity));
          println!("  Expires: {}", alert.expire_time);
          println!("More info: {}", alert.details_url);

          println!();
        }
      }
  }

  println!("{}", resp.current_weather.metadata.attribution_url); // Attribution labeling required by Apple
  
}
